core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -704,7 +705,27 @@ public int hashCode() {
         /**
          * Returns the next operation in the snapshot or <code>null</code> if we reached the end.
          */
-        Translog.Operation next() throws IOException;
+        Position next() throws IOException;
+
+    }
+
+    public static class Position {
```

Comment:
that's a trappy side-effect to have. I think we should add a `sizeInBytes()` to `Engine.Operation` then then engine is not required to walk on it's hands. I don't like that we have to change an interface to make this internal optimization work. I vote for fixing it on the `Engine.Operation` interface